### PR TITLE
Adding CloseChain operation to AMM

### PR DESCRIPTION
**Pull request by andresilva91**
## Motivation

This was raised during the weekly eng sync by @afck as a good follow up

## Proposal

Created the `CloseChain` operation, as well as added e2e tests for it, and tried to mimmick the Matching Engine's atomic swaps explanations here also

## Test Plan

CI

### CODE REVIEW

**Comment by andresilva91** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/linera-io/linera-protocol/1929?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#1929** <a href="https://app.graphite.dev/github/pr/linera-io/linera-protocol/1929?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#1928** <a href="https://app.graphite.dev/github/pr/linera-io/linera-protocol/1928?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1909** <a href="https://app.graphite.dev/github/pr/linera-io/linera-protocol/1909?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @andresilva91 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by andresilva91** [comment]
No idea why the `Check for outdated README.md` test is failing on this... I literally copy pasted from the `README.md` onto the `lib.rs` file like I always do, they're identical 🤔

File: examples/amm/src/contract.rs [hunk 1]
```diff
@@ -516,6 +525,30 @@ impl AmmContract {
...
+                        amount_token0,
+                        amount_token1,
+                    );
+                }
+
+                ensure!(
+                    *self.state.total_shares_supply.get() == Amount::ZERO,
+                    AmmError::UntrackedLiquidity
+                );
+                Ok(())
```
Comment (afck) [inline examples/amm/src/contract.rs:550]: You still need to call `runtime.close_chain()`.

Comment (andresilva91) [reply examples/amm/src/contract.rs:550]: 🤦🏻‍♂️ goot catch

File: examples/amm/src/state.rs [hunk 2]
```diff
@@ -1,11 +1,27 @@
...
 #[derive(RootView, async_graphql::SimpleObject)]
 #[view(context = "ViewStorageContext")]
 pub struct Amm {
-    // TODO(#968): We should support stateless Applications/empty user views
-    pub _dummy: RegisterView<u8>,
+    pub shares: MapView<Account, Amount>,
+    pub total_shares_supply: RegisterView<Amount>,
+}
+
+#[allow(dead_code)]
```
Comment (afck) [inline examples/amm/src/state.rs:17]: If it's only used in the contract, you could move this `impl` block into `contract.rs` instead.

File: examples/amm/src/contract.rs (hunk 1 above)
Comment (afck) [reply examples/amm/src/contract.rs:550]: Is this really resolved? I don't see it.

Comment (andresilva91) [reply examples/amm/src/contract.rs:550]: Sorry, I mark it as resolved when I have altered it locally, for me to keep track of the comments I'm addressing ha I'll update this with all the changes soon

**Pull request merged**